docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,300 @@
+If this ultimate timeout is reached without succesful key generation,
+the group dissolves and makes no further attempts.
+
+The intent of the two-phase retry protocol is
+to give node maintainers time to be alerted of and fix any problems
+that may have caused a large number of participants to drop out.
+
+==== Communication between participants
+
+Participants communicate using a broadcast channel.
+One-to-one messages are delivered by broadcasting them in encrypted form,
+using the key specific to the pair of participants.
+
+When operator Alice starts key generation,
+she establishes encryption keys with all other participants.
+When Alice sends a message that is intended for everyone,
+it is broadcast in plaintext to the entire group.
+When Alice sends a message intended only for Bob,
+she encrypts it using the key she has established with Bob,
+and broadcasts the ciphertext message.
+When Bob expects to receive an encrypted message from Alice,
+he takes Alice's broadcast message
+containing her encrypted payloads to all other participants,
+and decrypts the payload addressed to him
+using the key he has established with Alice.
+
+Both plaintext and ciphertext messages are authenticated by the sender.
+
+==== Removing inactive operators
+
+If Alice fails to receive the expected message from Bob
+within the specified timeout for the phase,
+she broadcasts this information to the rest of the group.
+If the majority of participants agree that Bob failed to send the message,
+Bob is added to the list of inactive members.
+
+If the majority does not agree that Bob failed to send the message,
+they will re-broadcast Bob's message to ensure that all members receive it.
+
+Whenever an operator is deemed to be inactive,
+the remaining participants start another key generation attempt
+with all inactive members removed.
+
+==== Dealing with misbehaviour
+
+The protocol in `tss-lib` is not attributable.
+When a participant is unable to proceed
+because they have received an invalid message from another participant,
+they will broadcast a special message declaring a _disqualification abort_
+and naming one or more participants who sent invalid messages.
+
+When a _disqualification abort message_ is broadcast by any member,
+group selection is aborted and no further attempts are made by the group.
+Instead, the group signs and submits an on-chain transaction
+announcing that key generation was aborted due to disqualification,
+and identifying the participants in the aborted attempt,
+the members who declared a disqualification abort
+and the members who were accused of sending invalid messages.
+All participants will also store their logs of the aborted attempt
+to facilitate later investigation.
+These logs must contain all messages sent and received by the participant,
+private keys the participant had established with all other participants,
+and all secret values the participant generated for the key generation attempt.
```

Comment:
I would probably start with a more lightweight approach: do not submit an on-chain transaction and just log a warning in logs of all clients. This needs to be resolved off-chain anyway.

Reply:
If there are on-chain repercussions I think we should have an on-chain signal, for both transparency (everyone can see when misbehaviour happens and there is no way to hide things behind the scenes) and the ability to limit the powers governance has (any governance-imposed punishments can only apply to operators named in an abort message signed by T members of a wallet)